add the option to disable functionalization in AOTDispatcher

bdhirsh opened the pull request:
I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164577
* #165372



cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

bdhirsh committed:
temp hacks to remove functionalization + get bitwise equivalence with aot_eager on llama3

[ghstack-poisoned]

bdhirsh commented:
the other change i needed was to tweak the test here: https://github.com/meta-pytorch/autoparallel/pull/176/files#diff-cbfe42e87df7867925a5e76ee407076c30f5ee141bf9a6785833679ef3c08533R59 so that it clones before the sum(), so we actually get contiguous tangents in the compiled backward (and don't need to clone)

ezyang commented on torch/_ops.py:
```diff
@@ -148,6 +148,8 @@ def inner(fn: Callable[_P, _T]) -> Callable[_P, _T]:
             )
 
+            if 'CompositeImplicit' in str(k) or 'Autograd' in str(k):
+                return fn
```
You're worrying about Python dispatcher, is that right? What if we don't enable Python dispatcher before running AOTAutograd so this code is all dead?

ezyang commented:
FYI this patch as is cannot run llama3 simple_fsdp on torchtitan:

```
  traceback : Traceback (most recent call last):
    File "/data/users/ezyang/b/pytorch/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 357, in wrapper
      return f(*args, **kwargs)
    File "/data/users/ezyang/b/torchtitan/torchtitan/train.py", line 596, in train
      self.train_step(data_iterator)
    File "/data/users/ezyang/b/torchtitan/torchtitan/train.py", line 496, in train_step
      loss = self.forward_backward_step(input_dict, labels)
    File "/data/users/ezyang/b/torchtitan/torchtitan/train.py", line 472, in forward_backward_step
      pred = model_parts[0](inputs, **extra_inputs)
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/eval_frame.py", line 418, in __call__
      return super().__call__(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/nn/modules/module.py", line 1777, in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/nn/modules/module.py", line 1788, in _call_impl
      return forward_call(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/eval_frame.py", line 886, in compile_wrapper
      return fn(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/nn/modules/module.py", line 1777, in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/nn/modules/module.py", line 1788, in _call_impl
      return forward_call(*args, **kwargs)
    File "/data/users/ezyang/b/torchtitan/torchtitan/models/llama3/model/model.py", line 394, in forward
      def forward(
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/eval_frame.py", line 1098, in _fn
      return fn(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/_functorch/aot_autograd.py", line 1134, in forward
      return compiled_fn(full_args)
    File "/data/users/ezyang/b/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 341, in runtime_wrapper
      all_outs = call_func_at_runtime_with_args(
    File "/data/users/ezyang/b/pytorch/torch/_functorch/_aot_autograd/utils.py", line 130, in call_func_at_runtime_with_args
      out = normalize_as_list(f(args))
    File "/data/users/ezyang/b/pytorch/torch/_functorch/_aot_autograd/utils.py", line 104, in g
      return f(*args)
    File "/data/users/ezyang/b/pytorch/torch/autograd/function.py", line 581, in apply
      return super().apply(*args, **kwargs)  # type: ignore[misc]
    File "/data/users/ezyang/b/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 2148, in forward
      fw_outs = call_func_at_runtime_with_args(
    File "/data/users/ezyang/b/pytorch/torch/_functorch/_aot_autograd/utils.py", line 130, in call_func_at_runtime_with_args
      out = normalize_as_list(f(args))
    File "/data/users/ezyang/b/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 529, in wrapper
      return compiled_fn(runtime_args)
    File "/data/users/ezyang/b/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 727, in inner_fn
      outs = compiled_fn(args)
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/backends/debugging.py", line 166, in run
      return forward_fn(args)
    File "/data/users/ezyang/b/pytorch/torch/fx/_lazy_graph_module.py", line 126, in _lazy_forward
      return self(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/fx/graph_module.py", line 843, in call_wrapped
      return self._wrapped_call(self, *args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/fx/graph_module.py", line 414, in __call__
      raise e
    File "/data/users/ezyang/b/pytorch/torch/fx/graph_module.py", line 401, in __call__
      return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
    File "/data/users/ezyang/b/pytorch/torch/nn/modules/module.py", line 1777, in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/nn/modules/module.py", line 1788, in _call_impl
      return forward_call(*args, **kwargs)
    File "<eval_with_key>.6", line 182, in forward
      graphsafe_run_with_rng_state = torch.ops.higher_order.graphsafe_run_with_rng_state(torch.ops.aten._scaled_dot_product_flash_attention.default, transpose_4, transpose_5, transpose_6, 0.0, True, scale = 0.25, rng_state = fwd_rng_state_0);  transpose_4 = transpose_5 = transpose_6 = fwd_rng_state_0 = None
    File "/data/users/ezyang/b/pytorch/torch/_prims/rng_prims.py", line 322, in __call__
      return super().__call__(op, *args, rng_state=rng_state, **kwargs)
    File "/data/users/ezyang/b/pytorch/torch/_ops.py", line 538, in __call__
      return wrapper()
    File "/data/users/ezyang/b/pytorch/torch/_ops.py", line 534, in wrapper
      return self.dispatch(
    File "/data/users/ezyang/b/pytorch/torch/_ops.py", line 507, in dispatch
      raise NotImplementedError(
  NotImplementedError: could not find kernel for HigherOrderOperator graphsafe_run_with_rng_state at dispatch key DispatchKey.AutogradCUDA (resolved from DispatchKey.AutogradCUDA)

============================================================
(b) [ezyang@devvm006.dkl0 ~/local/b/torchtitan] pp NGPU=1 CONFIG_FILE=./torchtitan/models/llama3/train_configs/debug_model.toml ./run_train.sh --compile.enable --compile.backend aot_eager --training.steps 10 --model.name simple_fsdp.llama3
```

ezyang commented:
WAR for this from Brian

```
diff --git a/torch/_ops.py b/torch/_ops.py
index fe8dd4fee62..c67eb35870b 100644
--- a/torch/_ops.py
+++ b/torch/_ops.py
@@ -148,8 +148,9 @@ class OperatorBase:
                 "Please register a mode for the DispatchKey.Python key instead."
             )

-            if 'CompositeImplicit' in str(k) or 'Autograd' in str(k):
-                return fn
+            if k == DispatchKey.CompositeImplicitAutograd or k == DispatchKey.Autograd:
+                if torch._C._dispatch_has_kernel(self.name()) and torch._C._dispatch_has_kernel_for_dispatch_key(self.name(), k):
+                    return fn
             if k in self.py_kernels:
                 raise RuntimeError(
                     f"Trying to override a python impl for {k} on operator {self.name()}"
```

bdhirsh committed:
Update on "temp hacks to remove functionalization + get bitwise equivalence with aot_eager on llama3"

Not for land. The right way for me to land this would be to clean up the "kill functionalization" changes into a separate PR backed by a config in AOTAutograd, and move the other changes into separate PRs




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

Not for land. The right way for me to land this would be to clean up the "kill functionalization" changes into a separate PR backed by a config in AOTAutograd, and move the other changes into separate PRs




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -296,13 +296,23 @@ def _has_tag_must_be_in_backward(node: fx.Node) -> bool:
+        isinstance(node.target, torch._ops.OpOverload)
+        and node.target._schema.is_mutable
     )
+    return _has_tag_is_backward(node) and is_mutable
```
The extra torture tests for mutation that I added in `test_aotdispatch.py` caught some more problems: we use `_must_be_in_forward/backward` in the default partitioner [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/partitioners.py#L206) when deciding how to splice out the fw and bw graphs. If we have a side-effectful op, then a mutation must be in the fw if it was traced at fw time (and vis versa). previously, this check only accounted for WithEffects ops and not arbitrary mutable ops.

ezyang replied:
This is very reasonable, but now I find myself a bit confused about how exactly the default partitioner works. Naively, I would have expected us to have just labeled all the forward nodes as forward and all the nodes traced during the backward() call as backwards, and then split this exactly. Since we don't do mutation in backwards, this would automatically cause all the mutations to go to forwards.

bdhirsh commented on torch/utils/checkpoint.py:
```diff
@@ -1187,7 +1187,7 @@ def unpack_hook_with_error_cb(holder):
     # Check if we are under AOTAutograd tracing
     # Checking that a functional mode is active should always do what we want
-    return torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.FUNCTIONAL) is not None
+    return torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.PROXY) is not None
```
added an AC test in `test_activation_checkpointing.py` for this

ezyang commented on torch/_functorch/_aot_autograd/graph_capture.py:
```diff
@@ -441,13 +463,14 @@ def aot_dispatch_autograd_graph(
 
     # There should be *NO* mutating ops in the graph at this point.
-    assert_functional_graph(fx_g.graph)
+    # /assert_functional_graph(fx_g.graph)
```
Gotta restore this under if branch

ezyang commented on torch/_functorch/_aot_autograd/graph_capture.py:
```diff
@@ -265,12 +280,13 @@ def aot_dispatch_base_graph(
+        fw_module.recompile()
+        copy_count2 = assert_functional_graph(fw_module.graph)
+        propagate_input_mutation_stacktraces(fw_module.graph)
+        assert copy_count == copy_count2
```
huh, intruiging

ezyang replied:
oh it was just hoisted up

ezyang commented on torch/_functorch/_aot_autograd/graph_capture.py:
```diff
@@ -441,13 +463,14 @@ def aot_dispatch_autograd_graph(
     torch._dynamo.utils.assert_no_fake_params_or_buffers(fx_g)
-    fx_g.graph.eliminate_dead_code()
+    if not aot_config.disable_functionalization:
+        fx_g.graph.eliminate_dead_code()
```
fwiw it OUGHT to be safe to DCE even when functionalization is off lol

ezyang commented on torch/_functorch/partitioners.py:
```diff
@@ -1016,11 +1026,50 @@ def default_partition(
+                            return True
+                elif u.target.is_view:
+                    tensor_arg_aliases.append(u)
+        return False
```
And so with a partitioner that exactly respects the original placement, I do not feel this function would be necessary.

bdhirsh replied:
Yeah this is very reasonable. If you would rather I rewrite the default partitioner completely to do this I'm happy to. Otherwise, I'll land this PR to get something shippable and we can do a broader refactor.

If we want to re-write the default partitioner this way, we'll have to think about the AC interaction (if we trace AC we get AC support "for free" if we do this strategy, but if we don't trace AC then the default partitioner will still need a graph pass to duplicate recompute in the backward, similar to what Jeffrey's PR does)

ezyang replied:
I basically want to do what Jeffreys PR does but in this pass lol

ezyang reviewed (approved):
I think it'll be good to merge this stuff so more people can kick the tires!

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -8194,24 +8302,32 @@ def test_aot_autograd_symbolic_exhaustive(self, device, dtype, op):
+        "test_aot_autograd_disable_functionalization_exhaustive",
         aot_autograd_failures,
     )
-    def test_aot_autograd_default_partition_exhaustive(self, device, dtype, op):
```
my plan in this PR is:

(1) have some tests that test "functionalization on, with min cut partitioner"

(2) have some tests that test "functionalization off, with default partitioner"

In the PR before this one, I wanted to test the default partitioner independently, so I added a test for "functionalization on, with default partitioner". My thought is that this test is unnecessary if we are testing the default partitioner with functionalization off, but happy to add it back

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "add the option to disable functionalization in AOTDispatcher"

I'm cleaning this PR up as a proper way of disabling functionalization via config in AOTDispatcher. I removed the non-functionalization related changes from the original version:

(1) preventing proxy mode (and functionalization) from incorrectly decomposing CIA ops (Ed has a PR for it here: https://github.com/pytorch/pytorch/pull/164939)

(2) preventing python-dispatcher-based decomps above autograd from running. I'm not doing this for now, will likely do it in a followup




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
